frame/support/src/storage/counted_map.rs

```diff
@@ -0,0 +1,314 @@
+	Counter: StorageValueT<u32>
+{
+	type Query = Map::Query;
+	type Hasher = Map::Hasher;
+	fn module_prefix() -> &'static [u8] {
+		Map::module_prefix()
+	}
+	fn storage_prefix() -> &'static [u8] {
+		Map::storage_prefix()
+	}
+	fn from_optional_value_to_query(v: Option<Value>) -> Self::Query {
+		Map::from_optional_value_to_query(v)
+	}
+	fn from_query_to_optional_value(v: Self::Query) -> Option<Value> {
+		Map::from_query_to_optional_value(v)
+	}
+}
+
+impl<Key, Value, Map, Counter>
+	StoragePrefixedMap<Value> for
+	CountedStorageMap<Key, Value, Map, Counter>
+where
+	Value: FullCodec,
+	Map: crate::StoragePrefixedMap<Value>,
+	Counter: StorageValueT<u32>
+{
+	fn module_prefix() -> &'static [u8] {
+		Map::module_prefix()
+	}
+	fn storage_prefix() -> &'static [u8] {
+		Map::storage_prefix()
+	}
+}
+
+impl<Key, Value, Map, Counter>
+	CountedStorageMap<Key, Value, Map, Counter>
+where
+	Key: FullCodec,
+	Value: FullCodec,
+	Map: StorageMapT<Key, Value> + crate::StoragePrefixedMap<Value>,
+	Counter: StorageValueT<u32, Query=u32>,
+{
+	// Internal helper function to track the counter as a value is mutated.
+	fn mutate_counter<
+		KeyArg: EncodeLike<Key> + Clone,
+		M: FnOnce(KeyArg, F) -> R,
+		F: FnOnce(I) -> R,
+		I,
+		R,
+	>(m: M, key: KeyArg, f: F) -> R {
+		let val_existed = Map::contains_key(key.clone());
+		let res = m(key.clone(), f);
+		let val_exists = Map::contains_key(key);
+
+		if val_existed && !val_exists {
+			// Value was deleted
+			Counter::mutate(|value| value.saturating_dec());
+		} else if !val_existed && val_exists {
+			// Value was added
+			Counter::mutate(|value| value.saturating_inc());
+		}
+
+		res
+	}
+
+	/// Get the storage key used to fetch a value corresponding to a specific key.
+	pub fn hashed_key_for<KeyArg: EncodeLike<Key>>(key: KeyArg) -> Vec<u8> {
+		Map::hashed_key_for(key)
+	}
+
+	/// Does the value (explicitly) exist in storage?
+	pub fn contains_key<KeyArg: EncodeLike<Key>>(key: KeyArg) -> bool {
+		Map::contains_key(key)
+	}
+
+	/// Load the value associated with the given key from the map.
+	pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> Map::Query {
+		Map::get(key)
+	}
+
+	/// Try to get the value for the given key from the map.
+	///
+	/// Returns `Ok` if it exists, `Err` if not.
+	pub fn try_get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> Result<Value, ()> {
+		Map::try_get(key)
+	}
+
+	/// Swap the values of two keys.
+	pub fn swap<KeyArg1: EncodeLike<Key>, KeyArg2: EncodeLike<Key>>(key1: KeyArg1, key2: KeyArg2) {
+		Map::swap(key1, key2)
+	}
+
+	/// Store a value to be associated with the given key from the map.
+	pub fn insert<KeyArg: EncodeLike<Key> + Clone, ValArg: EncodeLike<Value>>(key: KeyArg, val: ValArg) {
+		if !Map::contains_key(key.clone()) {
+			Counter::mutate(|value| value.saturating_inc());
+		}
+		Map::insert(key, val)
+	}
+
+	/// Remove the value under a key.
+	pub fn remove<KeyArg: EncodeLike<Key> + Clone>(key: KeyArg) {
+		if Map::contains_key(key.clone()) {
```

Comment:
```suggestion
		if Map::contains_key(&key) {
```

or use `mutate_exists`

Reply:
mutate_exists is less efficient here, because it queries and decode the value. I think we can't optimize remove further (except the duplication of computation of key, and clone here).

Reply:
the clone is fixed by using reference, it may end up of reference of refence though, maybe we should rely on the fact that the user gave the key in the most efficient way, so for big keys, it would be by reference and for small key like u32 by value.
In this case clone is more efficient than a reference.

Reply:
we can continue the discussion here if wanted https://github.com/paritytech/substrate/pull/9089#discussion_r651950997

Reply:
actually my thought are wrong, people can give a vec by value and wouldn't expect that it is cloned.